docs/develop/compiler.rst

```diff
@@ -0,0 +1,417 @@
+naming standard accordingly.
+
+.. note::
+   The order of these keywords in ``py/lexer.c`` should match the order of tokens in the enum
+   defined in ``py/lexer.h``.
+
+Parsing
+--------
+
+The parser takes the tokens produced by the lexer and converts them to an abstract syntax tree (AST) or
+*parse tree*. The implementation for the parser is defined in ``py/parse.c``. 
+
+The parser also maintains a table of constants for use in different aspects of parsing, similar to what a `symbol 
+table <https://steemit.com/programming/@drifter1/writing-a-simple-compiler-on-my-own-symbol-table-basic-structure>`_ 
+does.
+
+Several optimizations like `constant folding <http://compileroptimizations.com/category/constant_folding.htm>`_ 
+on integers for all operations i.e logical, binary, unary, etc, optimizing enhancements on parenthesis
+around expressions are performed during this phase and optimizations on strings.
+
+It's worth noting that *docstrings* are discarded and not accessible to the compiler.
+Even optimizations like `string interning <https://en.wikipedia.org/wiki/String_interning>`_ are 
+not applied to *docstrings*.
+
+Compiler passes
+---------------
+
+Like many compilers, MicroPython compiles all code to MicroPython bytecode or native code. The functionality
+that achieves this is implemented in ``py/compile.c``. The most relevant method you should know about is this:
+
+.. code-block:: c
+
+   mp_raw_code_t *mp_compile_to_raw_code(mp_parse_tree_t *parse_tree, 
+   qstr source_file, bool is_repl)
+
+The compiler compiles the code in several passes.
+
+First pass
+~~~~~~~~~~
+
+In the first pass, the compiler computes the stack sizes in scope:
+
+.. code-block:: c
+
+   // compile pass 1
+   comp->emit = emit_bc;
+   #if MICROPY_EMIT_NATIVE
+   comp->emit_method_table = &emit_bc_method_table;
+   #endif
+   uint max_num_labels = 0;
+   for (scope_t *s = comp->scope_head; s != NULL && comp->compile_error == MP_OBJ_NULL; s = s->next) {
+        #if MICROPY_EMIT_INLINE_ASM
+        if (s->emit_options == MP_EMIT_OPT_ASM) {
+            compile_scope_inline_asm(comp, s, MP_PASS_SCOPE);
+        } else
+        #endif
+        {
+            compile_scope(comp, s, MP_PASS_SCOPE);
+
+            // Check if any implicitly declared variables should be closed over
+            for (size_t i = 0; i < s->id_info_len; ++i) {
+                id_info_t *id = &s->id_info[i];
+                if (id->kind == ID_INFO_KIND_GLOBAL_IMPLICIT) {
+                    scope_check_to_close_over(s, id);
+                }
+            }
+        }
+   ..
+   }
+
+Other computations regarding scopes and identifiers are computed too.
+At this point, the number of labels that will be required in the emitted code is also determined and set.
+
+Second and third passes
+~~~~~~~~~~~~~~~~~~~~~~~
+
+The second and third passes involve computing the code size and emitting the ``inline assembler code`` for
```

Comment:
why is inline assembler code in quotes?

Reply:
This was meant for emphasis, but I can use italics instead.

Reply:
Ah, I see.  In that case italics would be better, because these quotes render to code-style.